File: models/cupynumeric/montecarlo.py
# python equivalent of gray-scott.jl to test the GC problem

import cupynumeric as np
from legate.timing import time
import csv
import sys 
import gc


def do_work(x, N):
    # time in same scope to avoid GC overhead
    gc.disable()
    start_time = time()
    res = (np.float32(10.0) / N) * np.sum(np.exp(-np.square(x)))
    end_time = time()
    gc.enable()
    return end_time - start_time

def mc_integration(N, n_steps, n_warmup):

    x = (np.float32(10.0)*np.random.rand(N)) - np.float32(5.0)

    for i in range(n_warmup):
        _ = do_work(x, N)

    times = []
    for i in range(n_steps):
        times.append(do_work(x, N))

    return np.sum(times)

def total_flops(N):
    return N    

gpus = int(sys.argv[1])
N = int(sys.argv[2])
n_samples = int(sys.argv[3])
warmup=2

print(f"[cuPyNumeric] Monte-Carlo Integration benchmark on {N} elements for {n_samples} iterations")

elapsed_time = mc_integration(N, n_samples, warmup)

total_time_ms = elapsed_time/1000.0
mean_time_ms = total_time_ms / n_samples
gflops = total_flops(N) / (mean_time_ms * 1e6) # GFLOP is 1e9

print(f"[cuPyNumeric] Mean Run Time: {mean_time_ms} ms")
print(f"[cuPyNumeric] FLOPS: {gflops} GFLOPS")


with open("./benchmarks/montecarlo/montecarlo.csv", "a", newline="") as csvfile:
    writer = csv.writer(csvfile)
    writer.writerow(["cuPyNumeric", gpus, N, 1, f"{mean_time_ms:.6f}", f"{gflops:.6f}"])

File: legacy/models/cunumeric/montecarlo.jl
using cuNumeric
using Legate
using LinearAlgebra
using Printf
using Statistics

get_time_us() = Legate.value(Legate.time_microseconds())

function initialize_cunumeric(N)
    A = (10.0f0 * cuNumeric.as_type(cuNumeric.rand(NDArray, N), Float32)) - 5.0f0
    GC.gc() # remove the intermediate FP64 arrays
    return A
end

function total_flops(N)
    return N 
end

function integrand(x)
    return exp(-square(x))
end

function do_work(x, N)
    return (10.0f0/N) * sum(integrand(x))
end

function mc_integration_cunumeric(N, n_samples, n_warmup)
    A = initialize_cunumeric(N)

    for idx in range(1,n_warmup)
        do_work(A, N)
        GC.gc()
    end

    times = []
    for idx in range(1, n_samples)
        GC.gc()
        t0 = get_time_us() # avoid timing GC as its not part of the algo
        res = do_work(A, N)
        t1 = get_time_us()
        push!(times, t1 - t0)
    end
    mean_time_ms = mean(times) / 1e3
    gflops = total_flops(N) / (mean_time_ms * 1e6) # GFLOP is 1e9

    return mean_time_ms, gflops
end

gpus = parse(Int, ARGS[1])
N = parse(Int, ARGS[2])
n_samples = parse(Int, ARGS[3])
n_warmup = 2

println("[cuNumeric.jl]  Monte-Carlo Integration benchmark on $(N) elements for $(n_samples) iterations, $(n_warmup) warmups")

mean_time_ms, gflops = mc_integration_cunumeric(N, n_samples, n_warmup)

println("[cuNumeric.jl]  Mean Run Time: $(mean_time_ms) ms")
println("[cuNumeric.jl]  FLOPS: $(gflops) GFLOPS")

open("./benchmarks/montecarlo/montecarlo.csv", "a") do io
    @printf(io, "%s,%d,%d,%d,%.6f,%.6f\n", "cuNumeric.jl", gpus, N, 1, mean_time_ms, gflops)
end
